src/adr/0034-mailbox-api-visibility-and-usage.md

```diff
@@ -0,0 +1,39 @@
+# 34. Mailbox API visibility and usage
+
+Date: 2020-04-27
+
+## Status
+
+Accepted (lazy consensus)
+
+## Context
+
+For some reasons, all mailboxes implementations rely on some abstract classes located in `mailbox-store` module (like 
+`StoreMailboxManager` and the `Mapper`s classes). For example, a `CassandraMailboxmanager` has to extend `StoreMailboxManager` 
+(that implements `Mailboxmanager` from the `mailbox-api`) that requires the implementation of some `Mapper`s.
+
+`Mapper`s are designed to provide low-level functions and methods on mailboxes. It's recurrent that we are tempted in 
+James, outside of the `mailbox` modules, to rely on some of those classes located in `mailbox-store` to have an easier 
+access on some user's mailboxes or messages. 
+
+## Decision
+
+We should never rely on classes defined in `mailbox-store` outside of the `mailbox` modules. The right way would be to 
+always rely on the `Manager`s defined in `mailbox-api` module to access mailboxes and messages.
+
+We should ensure the correctness of `Manager`s implementations by providing contract tests and not by sharing abstract 
+classes.
+
+## Consequences
+
+The issue with this design is that in this day, we can't seem to be able to act on a single message without knowing a 
```

Comment:
Sorry but I don't understand what it means.
Manager layer is responsible to ensuring access control on entities it manages.
A Message should be accessed only by either a User having access to it or to a technical component that has specific rights.
How is this problem related to the design describe previously?
Are you referring to the fact that by-passing access control is very handy? Because it's not stated explicitly in the Context.

Reply:
That was the all problem of this PR, starting the discussion: https://github.com/linagora/james-project/pull/3035, where I had to use the MailboxMapper outside of mailbox modules to be able to recompute the fast view projection of a single message of an user, as there was no other way at the moment. 

The conclusion, if I'm not wrong, was to bypass indeed the access control with the session, which would require a refactoring of this one. As the session refactoring should follow in an other ADR, I was not sure how much I should tell in this one. 

You are saying the context should state more that sometimes we need to bypass that access-control to access a message elsewhere?

Reply:
Currently, right enforcement requires a session to be open with the owner username for interacting with the manager layer, which is unknown given a simple `mailboxId`.

Reply:
Yes, I'm saying that, even if previous discussion is enough for me to know what you mean, it deserves to be written in Context part.